File: attendance/attendance/report/employee_attendance_summary/employee_attendance_summary.js
// Copyright (c) 2022, Peter Maged and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Employee Attendance Summary"] = {
  filters: [
    {
      fieldname: "company",
      label: __("Company"),
      fieldtype: "Link",
      options: "Company",
      default: frappe.defaults.get_user_default("Company"),
      reqd: 1,
    },
    {
      fieldname: "attendance_calculation",
      label: __("Attendance Calculation"),
      fieldtype: "Link",
      options: "Attendance Calculation",
      reqd: 1,
      on_change: function (query_report) {
        var attendance_calculation =
          query_report.get_values().attendance_calculation;
        if (!attendance_calculation) {
          return;
        }
        frappe.model.with_doc(
          "Attendance Calculation",
          attendance_calculation,
          function (r) {
            var fy = frappe.model.get_doc(
              "Attendance Calculation",
              attendance_calculation
            );
            frappe.query_report.set_filter_value({
              from_date: fy.payroll_start_date,
              to_date: fy.payroll_end_date,
            });
          }
        );
      },
    },
    {
      fieldname: "from_date",
      label: __("From Date"),
      fieldtype: "Date",
      read_only: 1,
      // reqd: 1,
      // default: new Date(new Date().getFullYear(), new Date().getMonth(), 1),
    },
    {
      fieldname: "to_date",
      label: __("To Date"),
      fieldtype: "Date",
      read_only: 1,
      // reqd: 1,
      // default: new Date(new Date().getFullYear(), new Date().getMonth() + 1, 0),
    },
    {
      fieldname: "employee",
      label: __("Employee"),
      fieldtype: "Link",
      options: "Employee",
      get_query: () => {
        var company = frappe.query_report.get_filter_value("company");
        return {
          filters: {
            company: company,
          },
        };
      },
    },

    {
      fieldname: "branch",
      label: __("Branch"),
      fieldtype: "Link",
      options: "Branch",
    },

    {
      fieldname: "department",
      label: __("Department"),
      fieldtype: "Link",
      options: "Department",
    },

    {
      fieldname: "grade",
      label: __("Grade"),
      fieldtype: "Link",
      options: "Grade",
    },
    {
      fieldname: "designation",
      label: __("Designation"),
      fieldtype: "Link",
      options: "Designation",
    },
  ],
};

File: attendance/attendance/report/employee_attendance_summary/employee_attendance_summary.py
# Copyright (c) 2022, Peter MAged and contributors
# For license information, please see license.txt

from unittest import result
import frappe
from frappe import _


def execute(filters=None):
    columns, data = [], []
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    columns = [
        {
            "fieldname": "employee",
            "label": _("Employee"),
            "fieldtype": "Link",
            "options": "Employee",
            "width": 250
        },
        {
            "fieldname": "employee_name",
            "label": _("Employee Name"),
            "fieldtype": "Data",
            "width": 150
        },
        # {
        #     "fieldname": "department",
        #     "label": _("Department"),
        #     "fieldtype": "Link",
        #     "options": "Department",
        #     "width": 120
        # },
        # {
        #     "fieldname": "grade",
        #     "label": _("Grade"),
        #     "fieldtype": "Link",
        #     "options": "Employee Grade",
        #     "width": 120
        # },
        # {
        #     "fieldname": "branch",
        #     "label": _("Branch"),
        #     "fieldtype": "Link",
        #     "options": "Branch",
        #     "width": 120
        # },
        # {
        #     "fieldname": "designation",
        #     "label": _("Designation"),
        #     "fieldtype": "Link",
        #     "options": "Designation",
        #     "width": 120
        # },
        
        {
            "fieldname": "total_present",
            "label": _("Present"),
            "fieldtype": "Float",
            "precision":2,
            "width": 90
        },
        
        {
            "fieldname": "total_absent",
            "label": _("Absent"),
            "fieldtype": "Float",
            "precision":2,
            "width": 90
        },
        
        {
            "fieldname": "total_wfh",
            "label": _("WFH"),
            "fieldtype": "Float",
            "precision":2,
            "width": 90
        },
        
        {
            "fieldname": "total_leave",
            "label": _("Leaves"),
            "fieldtype": "Float",
            "precision":2,
            "width": 90
        },
        
        {
            "fieldname": "shift_hours",
            "label": _("Target Hours"),
            "fieldtype": "Float",
            "precision":2,
            "width": 110
        },
        
        {
            "fieldname": "working_hours",
            "label": _("Working Hours"),
            "fieldtype": "Float",
            "precision":2,
            "width": 120
        },
        {
            "fieldname": "late_in",
            "label": _("Late"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "less_time",
            "label": _("Less"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "overtime",
            "label": _("Overtime"),
            "fieldtype": "Data",
            "precision":2,
            "width": 120
        },
        {
            "fieldname": "fingerprint",
            "label": _("Forget Fingerprint"),
            "fieldtype": "Float",
            "precision":2,
            "width": 150
        },
        {
            "fieldname": "permissions",
            "label": _("Permissions"),
            "fieldtype": "Float",
            "precision":2,
            "width": 120
        },

    ]

    return columns


def get_data(filters):
	conditions = get_employee_filters(filters or {})
	sql = f"""
			select 
					sum(case when log.holiday = 1 and log.status = 'On Leave' then 1 ELSE 0 END) as total_holiday,
						
					sum(case when  log.status = 'Absent' then 1 ELSE 0 END) as total_absent,
					
					sum(case when  log.status = 'Work From Home' then 1 ELSE 0 END) as total_wfh,
					
					sum(case when log.holiday <> 1 and log.status = 'On Leave'   then 1
					when log.holiday <> 1 and log.status = 'Half Day'   then 0.5
					ELSE 0 END) as total_leave,
					
					sum(case when log.status = 'Present'   then 1
					when log.status = 'Half Day'   then 0.5
					ELSE 0 END) as total_present,
					
					
					SUM(ifnull(log.shift_hours,0)) as shift_hours,
					
     				sum(log.working_hours) as working_hours,
					
					                    
					ifnull(sec_to_time(SUM(time_to_sec(log.less_time))),SEC_TO_TIME(0) ) as less_time,
					ifnull(sec_to_time(SUM(time_to_sec(log.late_in))),SEC_TO_TIME(0) ) as late_in,
					ifnull(sec_to_time(SUM(time_to_sec(log.overtime))),SEC_TO_TIME(0) ) as overtime,
								
					ifnull(sec_to_time(SUM(ifnull(log.overtime_factor*60,0))),SEC_TO_TIME(0) ) as overtime_factor,
					ifnull(sec_to_time(SUM(ifnull(log.late_factor*60,0))),SEC_TO_TIME(0) ) as late_factor,
					ifnull(sec_to_time(SUM(ifnull(log.less_time_factor*60,0))),SEC_TO_TIME(0) ) as less_time_factor,
								
						
					 SUM(CASE when ifnull(log.permission,'') <> '' then 1 else 0 END) as permissions ,
					 SUM(ifnull(log.forget_fingerprint,0)) as fingerprint ,
					 SUM(case when log.forget_fingerprint <> 0 then log.fingerprint_factor else 0 end) as fingerprint_factor ,
					
						
						
					log.employee,
					emp.employee_name,
					emp.company,
					emp.branch,
					emp.grade,
					emp.designation,
					emp.department

			from tabEmployee emp 
			Inner join tabAttendance  log on log.employee  = emp.name 
			where log.docstatus = 1 
			{conditions}
			GROUP by emp.name  			
			ORDER by emp.name ASC 

	"""
	# frappe.throw(sql)
	result = frappe.db.sql(sql,as_dict=1)
	data = result
	return data

def get_employee_filters(filters):
    from_date, to_date = filters.get('from_date'), filters.get('to_date')

    conditions = f" and log.attendance_date Between date('{from_date}') And date('{to_date}') "
    data = filters.get("company")
    if data:
        conditions += f" and emp.company = '{data}' "

    data = filters.get("employee")
    if data:
        conditions += f" and emp.name = '{data}' "

    data = filters.get("branch")
    if data:
        conditions += f" and emp.branch = '{data}' "
    data = filters.get("grade")
    if data:
        conditions += f" and emp.grade = '{data}' "
    data = filters.get("designation")
    if data:
        conditions += f" and emp.designation = '{data}' "
    data = filters.get("department")
    if data:
        conditions += f" and emp.department = '{data}' "
    return conditions
